### DISCUSSION TITLE
Derive two type classes with same unqualified name

### DISCUSSION THREAD

**Discussion by bishabosha** [scala/scala3#17892] [Feature Requests]
It is unfortunate that the path of a type class is not taken into consideration when creating cached fields for derived type class instances, leading to name collisions when two type classes share the same name, but have a different prefix.

## Minimized code

```Scala
object Foo {
  trait Sealed[A]
  object Sealed {
    def derived[A]: Sealed[A] = ???
  }
}

object Bar {
  trait Sealed[A]
  object Sealed {
    def derived[A]: Sealed[A] = ???
  }
}

enum Qux derives Foo.Sealed, Bar.Sealed {
  case A, B, C
}
```

## Output

```scala
-- Error: sandbox/mirror/example.scala:15:33 -----------------------------------
15 |enum Qux derives Foo.Sealed, Bar.Sealed {
   |                             ^^^^^^^^^^
   |                             duplicate type class derivation for Sealed
1 error found
```

## Expectation

the two derivation clauses refer to separate qualified types, they should be able to both be derived.

- at the location where this error is reported, the names in the deriving clause are already typed, (and we could use unique names for cached instances)
- this restriction is not really safe because two types in the deriving clause may be aliases

The following is a workaround - define a type alias and wrapper object to forward the `deriving` hook:

```scala
object Foo {
  trait Sealed[A]
  object Sealed {
    def derived[A]: Sealed[A] = ???
  }
}

object Bar {
  trait Sealed[A]
  object Sealed {
    def derived[A]: Sealed[A] = ???
  }
}

type BarSealed[A] = Bar.Sealed[A]
object BarSealed {
  inline def derived[A]: BarSealed[A] = Bar.Sealed.derived[A]
}

enum Qux derives Foo.Sealed, BarSealed {
  case A, B, C
}
```

**Comment by odersky** [comment]
That was intended exactly the way it is. We do not want name mangling here, so we fall back on the type alias trick. If this should be changed (which I would be skeptical about), it should be a feature request.

**Comment by rssh** [comment]
Related:  https://github.com/lampepfl/dotty/issues/11175

**Comment by bishabosha** [comment]
what about using underscores for the prefix? (less likely to collide but still possible)
`derived$Foo_Sealed`, `derived$Bar_Sealed`
